common/persistence/sql/sqlExecutionManager.go

```diff
@@ -469,42 +501,86 @@ func (m *sqlExecutionManager) Close() {
 	}
 }
 
-func (m *sqlExecutionManager) CreateWorkflowExecution(request *p.CreateWorkflowExecutionRequest) (*p.CreateWorkflowExecutionResponse, error) {
+func (m *sqlExecutionManager) CreateWorkflowExecution(request *p.CreateWorkflowExecutionRequest) (response *p.CreateWorkflowExecutionResponse, err error) {
+	err = m.txExecute("CreateWorkflowExecution", func(tx *sqlx.Tx) error {
+		response, err = m.createWorkflowExecutionTx(tx, request)
+		return err
+	})
+	return
+}
+
+func (m *sqlExecutionManager) createWorkflowExecutionTx(tx *sqlx.Tx, request *p.CreateWorkflowExecutionRequest) (*p.CreateWorkflowExecutionResponse, error) {
+	if request.CreateWorkflowMode == p.CreateWorkflowModeContinueAsNew {
+		return nil, &workflow.InternalServiceError{
+			Message: "CreateWorkflowExecution operation failed. Invalid CreateWorkflowModeContinueAsNew is used",
+		}
+	}
 	/*
 		(x) make a transaction
 		( ) check for a parent
 		(x) check for ContinueAsNew, update the current exec or create one for this new workflow
 		(x) create a workflow with/without cross dc
 	*/
-
-	tx, err := m.db.Beginx()
-	if err != nil {
-		return nil, &workflow.InternalServiceError{
-			Message: fmt.Sprintf("CreateWorkflowExecution failed. Failed to start transaction. Error: %v", err),
-		}
+	var row *currentExecutionRow
+	var err error
+	workflowID := *request.Execution.WorkflowId
+	if row, err = getCurrentExecutionIfExists(tx, int64(m.shardID), request.DomainID, workflowID); err != nil {
```

Comment:
Given that this is a locking read (and we do rangeID check after this), I am not sure if you still need the lockShard() towards the end of this method. I am not even sure if the locks acquired during a tx are re-entrant. So, this double acquisition of lock may not even work.

Reply:
These are for different purpose. Current execution is locks only a current_execution row for the workflowID. The shard is locked to make sure that the whole shard lease wasn't lost. This record is shared by all workflows in the shard and that's why it is locked at the end of the transaction to minimize contention.

Reply:
**> Current execution is locks only a current_execution row for the workflowID**
This is no longer true after your code change. lockCurrentExecutionIfExists does an INNER JOIN on shards table and so, it implicitly also locks the shard / rangeID. However, I verified that the locks acquired within a transaction are re-entrant and that's probably why the unit test succeed. Either ways, the lockShard() is no longer necessary AFAIK